server/team_matchmaking_service.py

```diff
@@ -0,0 +1,120 @@
+import time
+from typing import NamedTuple
+
+from .game_service import GameService
+from .decorators import with_logger
+from server.lobbyconnection import ClientError
+from .players import Player
+from .team_matchmaker.player_party import PlayerParty
+
+MapDescription = NamedTuple('Map', [("id", int), ("name", str), ("path", str)])
+GroupInvite = NamedTuple('GroupInvite', [("sender", Player), ("recipient", Player), ("party", PlayerParty), ("created_at", float)])
+
+PARTY_INVITE_TIMEOUT = 60 * 60 * 24  # secs
+
+@with_logger
+class TeamMatchmakingService:
+    """
+    Service responsible for managing the global team matchmaking. Does grouping, matchmaking, updates statistics, and
+    launches the games.
+    """
+
+
+    def __init__(self, games_service: GameService):
+        self.game_service = games_service
+        self.player_parties: dict[Player, PlayerParty] = dict()
+        self._pending_invites: dict[(Player, Player), GroupInvite] = dict()
+
+    def invite_player_to_party(self, sender: Player, recipient: Player):
+        if sender not in self.player_parties:
+            self.player_parties[sender] = PlayerParty(self, sender)
+
+        party = self.player_parties.get(sender)
+
+        if not party.owner == sender:
+            raise ClientError("You do not own this party.", recoverable=True)
+
+        if sender in recipient.foes:
+            raise ClientError("This person doesn't accept invites from you.", recoverable=True)
+
+        self._pending_invites[(sender, recipient)] = GroupInvite(sender, recipient, party, time.time())
+        recipient.lobby_connection.send({
+            "command": "party_invite",
+            "sender": sender.id
+        })
+
+    def accept_invite(self, recipient: Player, sender: Player):
+        if recipient not in self._pending_invites:
+            raise ClientError("You're not invited to a party", recoverable=True)
+
+        pending_invite = self._pending_invites.get((sender, recipient))
```

Comment:
Maybe this should be combined with the line further down:
```python
if recipient in self._pending_invites:
    self._pending_invites.pop((sender, recipient))
```

Since right now if `(sender, recipient)` doesn't exist in the dict, `get()` will return `None` and you will get an `AttributeError` for `pending_invite.sender` on the following line.

Reply:
Oh, just noticed you check for this existence on the previous line. So maybe you could do something like this:
```python
pending_invite = self._pending_invites.pop((sender, recipient), None)
if pending_invite is None:
    raise ClientError(...)
```

Or maybe this is better:
```python
if recipient not in self._pending_invites:
     raise ClientError("You're not invited to a party", recoverable=True)

pending_invite = self._pending_invites.pop((sender, recipient))
```

And then you don't need the extra existence check at the bottom

Reply:
which is why I check if (sender, recipient) is in self._pending_invites 2 lines above. Which I actually forgot to update when switching from recipient key to tuple key, so good you catched that

Reply:
Hmm I see some cases where those suggestions might not work correctly. I guess maybe you just replace the `.get()` with a bracket operator, since you already check for existence so it will never raise a `KeyError` anyways

Reply:
the extra existence check is void anyways :)

Reply:
Which cases do you see, I just removed it? If the party doesn't exist, I want to remove the invite anyways, and checking if the sender sent the invite is also void due to the sender now being part of the key of the map.

Reply:
Looks good to me now